mixer/pkg/runtime2/runtime.go

```diff
@@ -0,0 +1,195 @@
+		handlers:               handler.Empty(),
+		dispatcher:             dispatcher.New(identityAttribute, executorPool, enableTracing),
+		handlerPool:            handlerPool,
+
+		store: s,
+	}
+
+	// Make sure we have a stable state.
+	runtime.processNewConfig()
+
+	return runtime
+}
+
+// Dispatcher returns the runtime.Dispatcher that is implemented by this runtime package.
+func (c *Runtime) Dispatcher() runtime.Dispatcher {
+	return c.dispatcher
+}
+
+// StartListening directs Runtime to start listening to configuration changes. As config changes, runtime processes
+// the confguration and creates a dispatcher.
+func (c *Runtime) StartListening() error {
+	c.stateLock.Lock()
+	defer c.stateLock.Unlock()
+
+	if c.shutdown != nil {
+		return errors.New("already listening")
+	}
+
+	kinds := config.KindMap(c.snapshot.Adapters, c.snapshot.Templates)
+
+	data, watchChan, err := startWatch(c.store, kinds)
+	if err != nil {
+		return err
+	}
+
+	c.ephemeral.SetState(data)
+	c.processNewConfig()
+
+	c.shutdown = make(chan struct{})
+	c.waitQuiesceListening.Add(1)
+	go func() {
+		watchChanges(watchChan, c.shutdown, c.onConfigChange)
+		c.waitQuiesceListening.Done()
+	}()
+
+	return nil
+}
+
+// StopListening directs Runtime to stop listening to configuration changes. It will not unload the current
+// configuration, or close the existing adapters.
+func (c *Runtime) StopListening() {
+	c.stateLock.Lock()
+	defer c.stateLock.Unlock()
+
+	if c.shutdown != nil {
+		c.shutdown <- struct{}{}
+		c.shutdown = nil
+		c.waitQuiesceListening.Wait()
+	}
+}
+
+func (c *Runtime) onConfigChange(events []*store.Event) {
+	for _, e := range events {
+		c.ephemeral.ApplyEvent(e)
+	}
+	c.processNewConfig()
+}
+
+func (c *Runtime) processNewConfig() {
+	newSnapshot := c.ephemeral.BuildSnapshot()
+
+	oldHandlers := c.handlers
+
+	newHandlers := handler.NewTable(oldHandlers, newSnapshot, c.handlerPool)
+
+	builder := compiled.NewBuilder(newSnapshot.Attributes)
+	newRoutes := routing.BuildTable(
+		newHandlers, newSnapshot, builder, c.defaultConfigNamespace, log.DebugEnabled())
+
+	oldContext := c.dispatcher.ChangeRoute(newRoutes)
+
+	c.handlers = newHandlers
+	c.snapshot = newSnapshot
+
+	log.Debugf("New routes in effect:\n%s", newRoutes)
+
+	cleanupHandlers(oldContext, oldHandlers, newHandlers, maxCleanupDuration)
+}
+
+// maxCleanupDuration is the maximum amount of time cleanup operation will wait
+// before resolver ref count does to 0. It will return after this duration without
+// calling Close() on handlers.
+var maxCleanupDuration = 10 * time.Second
+
+var cleanupSleepTime = 500 * time.Millisecond
+
+func cleanupHandlers(oldContext *dispatcher.RoutingContext, oldHandlers *handler.Table, currentHandlers *handler.Table, timeout time.Duration) error {
+	start := time.Now()
+	for {
+		rc := oldContext.GetRefs()
+		if rc > 0 {
+			if time.Since(start) > timeout {
+				return fmt.Errorf("unable to cleanup resolver in %v time. %d requests remain", timeout, rc)
```

Comment:
It might be useful to add the resolver ID, don't know if there is a way to use it for debugging.